### ISSUE TITLE
Sharing-aware traversal library

### ISSUE THREAD

**Issue by atomb** [GaloisInc/crucible#417]
In support for #414, it would be useful to have a general library for doing sharing-aware traversals of What4 expressions. A sharing-aware pretty-printing library could be built on top of such a traversal, but many other forms of traversal could benefit from being able to share some of the memo table boilerplate.

**Comment by robdockins** [comment]
Yeah, we should definitely provide a pre-baked traversal combinator that plumbs through the necessary memoization.

**Comment by robdockins** [comment]
There are already some functions that help with this: `idxCacheEval` and `idxCacheEval'`.  They are useful building blocks; but, they don't fully package up a memoized traversal.

**Comment by atomb** [comment]
I notice that these are in `IO` but could be generalized to any `MonadIO`. Is there any reason not to do that?

**Comment by robdockins** [comment]
I don't think there would be any reason not to.

**Comment by atomb** [comment]
Yep, that's basically true, plus removal of one `liftIO`.

**Comment by travitch** [comment]
Diagnosing another problem, Kevin noticed that this is also affecting at least `assertSafe` (via `treeToPredicate`) in Lang.Crucible.CFG.Extension.Safety.
